scipy/_lib/_numpy_compat.py

```diff
@@ -45,3 +49,236 @@ def _assert_warns(warning_class, func, *args, **kw):
+                        "always", category=cat, message=mess)
+                else:
+                    module_regex = mod.__name__.replace('.', '\.') + '$'
+                    warnings.filterwarnings(
+                        "always", category=cat, message=mess,
+                        module=module_regex)
+                    self._tmp_modules.add(mod)
+            warnings.showwarning = self._showwarning
+            self._clear_registries()
+
+            return self
+
+        def __exit__(self, *exc_info):
+            warnings.showwarning = self._orig_show
+            warnings.filters = self._filters
+            self._clear_registries()
+            self._entered = False
+            del self._orig_show
+            del self._filters
+
+        def _showwarning(self, message, category, filename, lineno,
+                         *args, **kwargs):
+            for cat, _, pattern, mod, rec in (
+                    self._suppressions + self._tmp_suppressions)[::-1]:
+                if (issubclass(category, cat) and
+                        pattern.match(message.args[0]) is not None):
+                    if mod is None:
+                        if rec:
+                            self.log.append(
+                                WarningMessage(message, category, filename,
+                                               lineno, **kwargs))
+                        # Message and category match, either recorded or ignored
+                        return
+                    # Use startswith, because warnings strips the c or o from
+                    # .pyc/.pyo files.
+                    elif mod.__file__.startswith(filename):
+                        # The message and module (filename) match
+                        if rec:
+                            self.log.append(
+                                WarningMessage(message, category, filename,
+                                               lineno, *args, **kwargs))
+                        # The warning was not suppressed.
+                        return
+
+            # There is no filter in place, so pass to the outside handler
+            # unless we should only pass it once
+            if self._forwarding_rule == "always":
+                self._orig_show(message, category, filename, lineno,
+                                *args, **kwargs)
+                return
+
+            if self._forwarding_rule == "once":
+                signature = (message.args, category)
+            elif self._forwarding_rule == "module":
+                signature = (message.args, category, filename)
+            elif self._forwarding_rule == "location":
+                signature = (message.args, category, filename, lineno)
+
+            if signature in self._forwarded:
+                return
+            self._forwarded.add(signature)
+            self._orig_show(message, category, filename, lineno, *args, **kwargs)
+
+        def decorate(self, func):
+            """
```

Comment:
I am also not too happy with the name, since filter/suppression is now also recording. In some sense it is just catch_warnings pimped up ;).